torch/ao/quantization/pt2e/port_metadata_pass.py

```diff
@@ -88,12 +88,11 @@ def _port_metadata_for_input_quant_nodes(
         scale_node = choose_qparam_users.pop()
         dynamic_q_node = next(iter(scale_node.users.keys()))
         dynamic_q_node_users = _filter_sym_size_users(dynamic_q_node)
-        if len(dynamic_q_node_users) > 1:
-            raise InternalError(f"Expecting single user for {dynamic_q_node}")
-        dynamic_dq_node = dynamic_q_node_users.pop()
         _add_metadata(choose_qparams_node, node)
         _add_metadata(dynamic_q_node, node)
-        _add_metadata(dynamic_dq_node, node)
+        # q_node may have multiple users, e.g., in self-attention
+        for dynamic_dq_node in dynamic_q_node_users:
```

Comment:
this is a bit weird, why does q node have multiple users for dynamic quantization? doesn't this break the `choose_qparams -> q -> dq` pattern?

Reply:
Hi @jerryzh168. In the QKV case, multiple `qd`'s are added by the `DuplicateDQPass` here: https://github.com/pytorch/pytorch/blob/9dda4b20a0dcc4786a130c12fc3e4a6a1ecf23d8/torch/ao/quantization/quantize_pt2e.py#L235C1-L238C1
I think this affects both static quant and dynamic quant. Do you have any suggestions on how to deal with this? Thanks

Reply:
oh I see, I think in this case we'd need to duplicate not just dq, but also the entire chain of "choose_qparams -> q -> dq". this sounds like a hacky solution though, maybe we want to redesign this. cc @kimishpatel

Reply:
I think it's also related to how we match the patterns, if we manually match the pattern this could work as well. see https://github.com/pytorch/pytorch/pull/83459

Reply:
Yes. For X86Inductor, we plan to match the `dq -> linear` pattern only. A single `choose_qparam -> q` with duplicate `dq`'s is what we want. If a backend wants to duplicate the entire chain of `choose_qparams -> q -> dq`, can this be done during lowering?

Reply:
Sorry, am not reading a whole lot so just asking my question. Is the issue that dq is duplicated even for dynamic quant case?

Reply:
>  single choose_qparam -> q with duplicate dq's is what we want

Why not the whole chain? That sounds a bit more reasonable to me.

Reply:
yeah I agree with @kimishpatel, I think matching the whole chain will be more robust, otherwise how do you know it's a pattern for dynamic quantization

Reply:
For X86Inductor backend, we do not need to know it's dynamic quantization actually. We only want to fuse `dq - linear` and leave `choose_qparam - q` in the graph. I have some explanation here: https://github.com/pytorch/pytorch/pull/115337/files#r1440093632
I will update this PR so that `dq` is not duplicated for dynamic quantization.

Reply:
Please review again. Thanks!

Reply:
@Xia-Weiwen btw this may become relevant for some of the works we are doing elsewhere. Specifdicaly in LLM kind of models dynamically quantizing q, k, v projection makes sense. which is what you are trying to do. In this case for xnnpack quantizer, we will have a similar issue. My earlier suggestion to not duplicate dq for dynamic quant was based on assumtion that we want to treat "dynamically quantized linear" differently from "quantized linear". But this discussion has made me think that you can argue that "dynamically quantized linear" = "dynamic quantized act" + "static quantized linear". If we follow this, then your changes of duplicating dq for dynamic quant still makes sense. I want to think a bit more on this. @jerryzh168 thoughts?

Reply:
Thanks for the comment. I will land this PR then.